File: 47.permutations-ii.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var nextPermutation = function(nums) {
    for (let i = nums.length - 1; i > 0; i--) {
        if (nums[i - 1] < nums[i]) {
            for (let j = nums.length - 1; j >= i; j--) {
                if (nums[j] > nums[i - 1]) {
                    let tmp = nums[i - 1];
                    nums[i - 1] = nums[j];
                    nums[j] = tmp;
                    for (let j = i, k = nums.length - 1; j < k; j++, k--) {
                        let tmp = nums[j];
                        nums[j] = nums[k];
                        nums[k] = tmp;
                    }
                    return;
                }
            }
        }
    }
    nums.reverse();
    return;
};

/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    let ans = [];
    ans.push(nums.slice());
    nextPermutation(nums);
    while (!_.isEqual(nums, ans[0])) {
        ans.push(nums.slice());
        nextPermutation(nums);
    }
    return ans;
};

File: 47.permutations-ii.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        for (int i = nums.size() - 1; i > 0; i--) {
            if (nums[i - 1] < nums[i]) {
                for (int j = nums.size() - 1; j >= i; j--) {
                    if (nums[j] > nums[i - 1]) {
                        int tmp = nums[i - 1];
                        nums[i - 1] = nums[j];
                        nums[j] = tmp;
                        reverse(nums.begin() + i, nums.end());
                        return;
                    }
                }
            }
        }
        reverse(nums.begin(), nums.end());
        return;
    }

    vector<vector<int>> permuteUnique(vector<int>& nums) {
        vector<vector<int>> ans;
        ans.push_back(nums);
        nextPermutation(nums);
        while (nums != ans[0]) {
            ans.push_back(nums);
            nextPermutation(nums);
        }
        return ans;
    }
};